### PR TITLE
Make sure dev ui only loads on localhost

### LINKED ISSUE #40828: dynamic default index page has bad behavior when access from non localhost
### Describe the bug

I'm using ngrok to expose my app as github/zulip bots.

the new default landing page then now just shows up with /q/dev-ui/welcome and stuck at "quarkus" logo.

This is due to default cors preventing such access.

Same will happen I assume in web ide dev environments etc.

### Expected behavior

that *some* index page can be shown with a hint of what went wrong - unauthorized access would be useful.

or at least *some* info.

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by phillip-kruger**
Fix #40828

### CODE REVIEW

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUICORSFilter.java [hunk 1]
```diff
@@ -46,18 +50,42 @@ public void handle(RoutingContext event) {
...
-                    || origin.startsWith(HTTP_LOCAL_HOST_IP) || origin.startsWith(HTTPS_LOCAL_HOST_IP)) {
+        if (hostIsValid(event)) {
+            if (origin == null) {
                 corsFilter().handle(event);
             } else {
-                LOG.errorf("Only localhost origin is allowed, but Origin header value is: %s", origin);
-                response.setStatusCode(403);
-                response.setStatusMessage("CORS Rejected - Invalid origin");
-                response.end();
+                if (origin.startsWith(HTTP_LOCAL_HOST) || origin.startsWith(HTTPS_LOCAL_HOST)
```
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUICORSFilter.java:57]: What happens if cors are disabled? 

I've been working on a setup where we need to disable the dev-ui CORS (I understand the risk).

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUICORSFilter.java [hunk 2]
```diff
@@ -46,18 +50,42 @@ public void handle(RoutingContext event) {
...
+        try {
+            URI uri = new URI(event.request().absoluteURI());
+            URL url = uri.toURL();
+            String host = url.getHost();
+
+            if (host.equals(LOCAL_HOST) || host.equals(LOCAL_HOST_IP)) {
+                return true;
             }
+        } catch (MalformedURLException | URISyntaxException e) {
+            e.printStackTrace();
```
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUICORSFilter.java:87]: Better use a logger.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUICORSFilter.java (hunk 1 above)
Comment (phillip-kruger) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUICORSFilter.java]: Ok I moved this logic to a dedicated filter

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public void handle(RoutingContext event) {
+        HttpServerResponse response = event.response();
+        if (hostIsValid(event)) {
+            event.next();
+        } else {
+            LOG.error("Dev UI: Only localhost is allowed");
+            response.setStatusCode(403);
+            response.setStatusMessage("Dev UI: Only localhost is allowed - Invalid host");
```
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:31]: I think there will be a problem with this.
Typically, the case I'm facing right now is that I need the dev UI to work from DevSpaces. Unfortunately, it's not localhost but an exposed route. 

We need to allow a list of allowed hosts.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpServerResponse;
+import io.vertx.ext.web.RoutingContext;
+
+public class LocalHostOnlyFilter implements Handler<RoutingContext> {
+    private static final Logger LOG = Logger.getLogger(LocalHostOnlyFilter.class);
+
+    private static final String LOCAL_HOST = "localhost";
+    private static final String LOCAL_HOST_IP = "127.0.0.1";
```
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:18]: Need to support IPv6 (`::1`)

**Review by cescoffier — APPROVED** [review]

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java (hunk 3 above)
Comment (phillip-kruger) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:31]: How about a config option where you can add allowed hosts ?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java (hunk 4 above)
Comment (phillip-kruger) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:18]: I can add that

Comment (maxandersen) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:18]: I can't help feel like we are somehow circumventing or not utilizing the right cors setup if we have to do these tricks? how come this is not caught by normal cors?

Comment (phillip-kruger) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:18]: This is not really CORS. It's allowed hosts. In the case if fails the origin header is null, so CORS can not work

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java (hunk 3 above)
Comment (phillip-kruger) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/LocalHostOnlyFilter.java:35]: You can now add hosts that should be allowed

**Comment by slallemand** [comment]
Hello @phillip-kruger 

Thanks for this improvement !
Is there a way to allow a domain with these settings ? 
I want to use the dev-ui in DevSpaces so I know my domain but not the FQDN of my Quarkus service.

I tried with *.domain , %.domain with no success.

Thanks !

**Comment by cescoffier** [comment]
@phillip-kruger, can you look at @slallemand's request?

**Comment by phillip-kruger** [comment]
@slallemand - so you want to be able to pass in a regex and if it match we should allow ? Will that work ?

**Comment by slallemand** [comment]
@phillip-kruger exactly !

**Pull request merged**